pkg/controller/ibmblockcsi/ibmblockcsi_controller.go

```diff
@@ -355,25 +350,23 @@ func (r *ReconcileIBMBlockCSI) updateStatus(instance *ibmblockcsi.IBMBlockCSI, o
 				logger.Error(err, "failed to get controller pod")
 				return err
 			}
-			if originalStatus.ControllerReady || !r.areAllPodImagesSynced(controllerStatefulset, controllerPod) {
-				logger.Info("controller requires restart",
-							"ReadyReplicas", controllerStatefulset.Status.ReadyReplicas,
-							"Replicas", controllerStatefulset.Status.Replicas)
-				controllerRestart = true
+	
+			for _, containerStatus := range controllerPod.Status.ContainerStatuses {
+				if containerStatus.State.Waiting != nil {
+					if containerStatus.State.Waiting.Reason == "ImagePullBackOff" {
```

Comment:
why is this needed?
I don't think we should restart the pod for each faulty container

Reply:
we still need to handle the problem with statefulsets that doesn't redoply pods in ImagePullBackOff even though the user updates their images

Reply:
oh, then you can still use `areAllPodImagesSynced` for that

Reply:
what is `areAllPodImagesSynced`? this is an existing function? or its a function that you want me to add?

Reply:
done

Reply:
I still don't see why we should restart the pod for each faulty container, not why we need to check the waiting reason.
can't we just keep the original code here (minus the `originalStatus` check)?

Reply:
We wanted to remove the original code and move it to the service account reconciler.
The original code was also restarting the controller pods for each faulty container, this code is restarting only when there is an ImagePullBackOff and the function `areAllPodImagesSynced` returns false so it minimizes the restating action to less events

Reply:
the original code did **not** restart the controller pod **for each** faulty container, it restarted the pod **once** if **any** container image was not in sync with the statefulset.
so current diff can potentially **add** restart events

Reply:
but the current diff also restart the pod once if any container image was not in sync with the statefulset, but when there is `ImagePullBackOff`, I can remove the if of the `ImagePullBackOff` if you like

Reply:
it seems to me as if the current diff restarts **for each** faulty container.
what am I missing here?

Reply:
I would remove the `ImagePullBackOff` regardless

Reply:
I will do it like you suggested at the beginning

Reply:
done